Add the ability to use mcumgr over udp.

bperseghetti opened the pull request.

bperseghetti committed:
Add MCUmgr UDP transport and device info script

Enable MCUmgr management over UDP for the optical_flow application,
allowing remote device queries without a serial connection.

Add Python script (scripts/mcuboot_device_info.py) that uses the SMP
protocol over UDP to retrieve device information:
- Bootloader info (MCUboot mode, version)
- OS/Application info (kernel, version, build date)
- Image state (slots, versions, hashes, flags)
- Task statistics
- MCUmgr parameters

Configuration added to prj.conf:
- MCUmgr with UDP transport on default port 1337
- OS management group (info, bootloader info, taskstat)
- Image management group (state, slot info)

Usage: python scripts/mcuboot_device_info.py 192.0.2.1

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>

bperseghetti commented:
@ubieda would you mind testing this on your system and seeing what type of performance you are getting?

Connect your mcxn-t1 to your machine with the t1-adapt and use/edit a netplan in `/etc/netplan` to set whatever iface you choose to something like:
```yaml
network:
  version: 2
  ethernets:
    <your-iface-name-of-choice>:
      renderer: NetworkManager
      match: {}
      addresses:
      - "192.0.2.2/24"
      networkmanager:
        uuid: "<your-already-existing-uuid>"
        name: "netplan-<your-iface-name-of-choice>"
        passthrough:
          ethernet.speed: "100"
          ipv4.address1: "192.0.2.2/24,255.255.255.0"
          ipv4.method: "manual"
          ipv6.method: "disabled"
          ipv6.ip6-privacy: "-1"
          proxy._: ""
```
Then `sudo netplan apply`

If you have a zeth network device (`ifconfig zeth`) it will collide with the default network config so remove it for now:
```bash
sudo ip link set zeth down
sudo ip link delete zeth
```
Note you might need to install `sudo apt-get install python3-cbor2`

Steps to test:
```bash
cd ~/cognipilot/cranium/src
git clone git@github.com:CogniPilot/spinali_optical_flow
git clone git@github.com:CogniPilot/pixart_msgs
git clone git@github.com:CogniPilot/afbr_msgs
cd synapse_ros
git fetch
git checkout pr-add-argus-of-imu-mag
cd ~/cognipilot/cranium
colcon build --symlink-install
source ~/.bashrc
ros2 launch spinali_optical_flow_integration argus_to_pointcloud.launch.py
```
In another terminal run:
```bash
cd ~/cognipilot/ws/spinali/
./scripts/mcuboot_device_info.py 192.0.2.1 -v
```
Should look like this:
<img width="1437" height="2470" alt="Screenshot from 2025-12-17 15-05-24" src="https://github.com/user-attachments/assets/7398eb55-20e7-48bc-85fa-c137c680d794" />

ubieda commented on scripts/mcuboot_device_info.py:
Why not just use https://github.com/intercreate/smpclient ?

bperseghetti replied:
Mostly I wanted this as a minimal encapsulated tool as much as possible as a reference for how to integrate some parts of these into eventually a Rust based HCDF manifest parser that can then read and set firmware in cyber physical system. This is not meant to be a long term tool, just a convenience script to get familiar with SMP and CBOR.

bperseghetti replied:
I also added a custom mcmgr hook that goes beyond the typical "client tools".

bperseghetti committed:
Add hardware ID support via MCUmgr custom hook

Add shared MCUmgr hooks module (drivers/synapse/mcumgr_hooks) that
extends the OS info command with a custom 'h' format character to
return the chip's unique hardware ID via the hwinfo driver.

Enable in applications with CONFIG_SPINALI_SYNAPSE_MCUMGR_HOOKS=y

Updates to mcuboot_device_info.py:
- Add hardware ID display from custom 'h' format
- Add MAC address lookup from ARP table
- Add build date/time to OS info output
- Improve OS info field parsing and formatting
- Remove unsupported MPSTAT command (rc=8 error)

Additional prj.conf changes:
- Enable thread names and monitoring for taskstat
- Enable build date/time in OS info
- Set network hostname to "optical-flow"

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>

bperseghetti commented:
My next goal is to use AMP between the two cores so this can run on the non-FPU CM33 with the rest of the enet stack.

bperseghetti commented:
Ooooooof, I guess maybe I spoke too soon, started digging into it further and don't think that would work well on CPU1 (the non-FPU CM33)... It only has 64 KB of RAM, which is not enough for the network stack and buffers and ENET peripheral is configured for CPU0 access only, so CPU1 cannot directly use the ethernet hardware. And if I were to switch it the network would be unavailable until CPU1 boots since CPU0 launches the secondary core from what I can seem to find..... 

That and any communication between cores would add IPC overhead and extra latency to every packet if it were to even fit (which it won't). 

@PetervdPerk-NXP thoughts on what we can do with second core? 64KB of RAM and 200KB flash sure does make it more limited for most of the applications I can think of offhand.

bperseghetti commented:
My update to add the hook for getting the Hardware ID(s) now returns this for instance:
<img width="736" height="1898" alt="image" src="https://github.com/user-attachments/assets/8dedf631-b012-4915-8bde-11a9a7621cb7" />

bperseghetti committed:
Update MCUboot build instructions and zephyr_boards

Update README with simplified MCUboot build command using cmake cache
file from zephyr_boards. Bump zephyr_boards to include MCUboot board
configuration for mr_mcxn_t1.

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>

bperseghetti committed:
west: Update zephyr_boards to fix MCUboot partition

Update zephyr_boards to include the fix for MCUboot being linked
to the wrong flash partition address on mr_mcxn_t1.

Also add note to README about running west blobs fetch after update.

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>

bperseghetti committed:
optical_flow: enable led-force-on for PAA3905

Enable the led-force-on property to work around inverted LED control
logic on this board.

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>

bperseghetti committed:
Test hcdf.

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>

bperseghetti committed:
Revert "optical_flow: enable led-force-on for PAA3905"

This reverts commit d792c27248c76f9e571fefc41f657374b7333c4b.

bperseghetti committed:
optical_flow: update HCDF to v2.0 with ports and sensors

- Update URL to new path structure (mr_mcxn_t1/optical-flow/optical-flow.hcdf)
- Add SHA config for HCDF cache validation (4579dee4)

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>

bperseghetti committed:
Add HCDF SHA validation and update HCDF URLs

- Add hcdf_models module to west.yml
- Enable build-time HCDF SHA validation in optical_flow
- Update optical_flow SHA to b05fb19d
- Fix rtk_gnss HCDF URL format and add SHA
- Update Kconfig help text for HCDF_URL and HCDF_SHA

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>
